dom.go

```diff
@@ -1,92 +1,164 @@
 package vecty
 
-import "github.com/gopherjs/gopherjs/js"
+import (
+	"fmt"
 
-// Component represents a Vecty component.
-type Component interface {
-	Markup
-	Reconcile(oldComp Component)
-	Node() *js.Object
-}
+	"github.com/gopherjs/gopherjs/js"
+)
 
-// Render renders a component into the given container element. It is appended
-// as a child element.
-func Render(comp Component, container *js.Object) {
-	comp.Reconcile(nil)
-	container.Call("appendChild", comp.Node())
+// Core implements the Context method of the Component interface, and is the
+// core/central struct which all Component implementations should embed.
+type Core struct {
+	prevRender *HTML
 }
 
-// RenderAsBody renders the given component as the body of the page, replacing
-// whatever existing content in the page body there may be.
-func RenderAsBody(comp Component) {
-	doc := js.Global.Get("document")
-	body := doc.Call("createElement", "body")
-	Render(comp, body)
-	if doc.Get("readyState").String() == "loading" {
-		doc.Call("addEventListener", "DOMContentLoaded", func() { // avoid duplicate body
-			doc.Set("body", body)
-		})
-		return
-	}
-	doc.Set("body", body)
-}
+// Context implements the Component interface.
+func (c *Core) Context() *Core { return c }
 
-type textComponent struct {
-	text string
-	node *js.Object
-}
+// Component represents a single visual component within an application. To
+// define a new component simply implement the Render method and embed the Core
+// struct:
+//
+// 	type MyComponent struct {
+// 		vecty.Core
+// 		... additional component fields (state or properties) ...
+// 	}
+//
+// 	func (c *MyComponent) Render() *vecty.HTML {
+// 		... rendering ...
+// 	}
+//
+type Component interface {
+	// Render is responsible for building HTML which represents the component.
+	Render() *HTML
 
-// Apply implements the Markup interface.
-func (s *textComponent) Apply(element *Element) {
-	element.Children = append(element.Children, s)
+	// Context returns the components context, which is used internally by
+	// Vecty in order to store the previous component render for diffing.
+	Context() *Core
 }
 
-func (s *textComponent) Reconcile(oldComp Component) {
-	if oldText, ok := oldComp.(*textComponent); ok {
-		s.node = oldText.node
-		if oldText.text != s.text {
-			s.node.Set("nodeValue", s.text)
-		}
-		return
-	}
+// ComponentOrHTML represents one of:
+//
+//  Component
+//  *HTML
+//
+// If the underlying value is not one of these types, the code handling the
+// value is expected to panic.
+type ComponentOrHTML interface{}
```

Comment:
Union-like types are used to differentiate between HTML (`DOMNode` or `TextNode`) and components (user built components). This way `*HTML` (previously `Element`) can distinguish the two and straightforward decisions about where the previous render is etc.